llap-tez/src/java/org/apache/hadoop/hive/llap/tezplugins/metrics/BlacklistingLlapMetricsListener.java

```diff
@@ -0,0 +1,200 @@
+          LOG.debug("Already too many blacklisted nodes. Skipping.");
+          return;
+        }
+      }
+
+      if (requestHandled > this.minServedTasksNumber) {
+        workerNum++;
+        sumAverageTime += averageTime;
+        if ( averageTime > maxAverageTime) {
+          maxAverageTime = averageTime;
+          maxAverageTimeEmptyExecutors = emptyExecutor;
+          maxAverageTimeMaxExecutors = maxExecutors;
+          maxAverageTimeIdentity = workerIdentity;
+        }
+        sumEmptyExecutors += emptyExecutor;
+      }
+    }
+
+    // If we do not have enough data then return.
+    if (workerNum < 2) {
+      return;
+    }
+
+    LOG.debug("Found slowest node {} with data: " +
+                  "sumAverageTime={}, " +
+                  "sumEmptyExecutors={}, " +
+                  "maxAverageTime={}, " +
+                  "maxAverageTimeEmptyExecutors={}, " +
+                  "maxAverageTimeMaxExecutors={}, " +
+                  "workerNum={}, " +
+                  "maxAverageTimeIdentity={}, " +
+                  "blacklistedNodes={}",
+        sumAverageTime, sumEmptyExecutors, maxAverageTime, maxAverageTimeEmptyExecutors,
+        maxAverageTimeMaxExecutors, workerNum, maxAverageTimeIdentity, blacklistedNodes);
+    // Check if the slowest node is at least timeThreshold times slower than the average
+    long averageTimeWithoutSlowest = (sumAverageTime - maxAverageTime) / (workerNum - 1);
+    if (averageTimeWithoutSlowest * this.timeThreshold < maxAverageTime) {
+      // We have a candidate, let's see if we have enough empty executors.
+      long emptyExecutorsWithoutSlowest = sumEmptyExecutors - maxAverageTimeEmptyExecutors;
+      if (emptyExecutorsWithoutSlowest > maxAverageTimeMaxExecutors * this.emptyExecutorsThreshold) {
+        // Seems like a good candidate, let's try to blacklist
+        try {
+          LOG.debug("Trying to blacklist node: " + maxAverageTimeIdentity);
+          setCapacity(maxAverageTimeIdentity, 0, 0);
+        } catch (Throwable t) {
+          LOG.debug("Can not blacklist node: " + maxAverageTimeIdentity, t);
+        }
+      }
+    }
+  }
+
+  protected void setCapacity(String workerIdentity, int newExecutorNum, int newWaitQueueSize)
+      throws IOException, ServiceException {
+    long currentTime = System.currentTimeMillis();
+    if (currentTime > nextCheckTime) {
+      LlapZookeeperRegistryImpl.ConfigChangeLockResult lockResult =
+          registry.lockForConfigChange(currentTime + this.minConfigChangeDelayMs);
+
+      LOG.debug("Got result for lock check: {}", lockResult);
+      if (lockResult.isSuccess()) {
+        LOG.info("Setting capacity for workerIdentity={} to newExecutorNum={}, newWaitQueueSize={}",
+            workerIdentity, newExecutorNum, newWaitQueueSize);
+        LlapServiceInstance serviceInstance = registry.getInstances().getInstance(workerIdentity);
+        LlapManagementProtocolClientImpl client = clientFactory.create(serviceInstance);
```

Comment:
Would it make sense to have one client implementation as member variable instead of creating new ones here? This one instance could still be created lazy...

Reply:
This client implementation is for a specific worker. I hope we do not have to reuse the client to blacklist the same node twice :D